### PR TITLE
Add support for model after commit hooks

### PR DESCRIPTION
**Pull request by notbakaneko**
Primitive `after_commit` support to be used by elasticsearch updates on model save when transaction completes.

### CODE REVIEW

**Comment by nanaya** [comment]
Can use an example on how to use this ಠ_ಠ

**Comment by notbakaneko** [comment]
AfterCommitTest :trollface:

File: app/Models/Model.php [hunk 1]
```diff
@@ -93,4 +112,23 @@ public function saveOrExplode($options = [])
...
+
+            return;
+        }
+
+        if ($this instanceof AfterCommit) {
+            $transaction->addCommittable($this);
+        }
+
+        if ($this instanceof AfterRollback) {
+            $this->addRollbackable($this);
```
Comment (nanaya) [inline app/Models/Model.php]: `$transaction`?

File: app/Providers/TransactionStateServiceProvider.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+     * @return void
+     */
+    public function register()
+    {
+        // Connection after commit tracker.
+
+        // Laravel uses a singleton DatabaseManager that reuses connections
+        // by name, so a singleton for tracking should be good enough.
+        // The alternative is to override Connection and its subclasses.
+        $this->app->singleton('TransactionState', function ($app) {
```
Comment (nanaya) [inline app/Providers/TransactionStateServiceProvider.php]: there's `TransactionState` class and then there's `TransactionState` singleton which is actually `TransactionStateManager` ಠ_ಠ

File: app/Models/Model.php [hunk 3]
```diff
@@ -21,6 +21,9 @@
 namespace App\Models;
 
 use App\Exceptions\ModelNotSavedException;
+use App\Libraries\Transactions\AfterCommit;
```
Comment (nanaya) [inline app/Models/Model.php]: I'd just add an empty method in base model and let the actual models override it instead of having to do the ceremonies...

Comment (notbakaneko) [reply app/Models/Model.php]: Don't want to be tracking models that don't do anything for after commit

**Comment by nanaya** [comment]
still randomly failing =[

**Comment by notbakaneko** [comment]
:rage3:

**Comment by notbakaneko** [comment]
all the reverts :trollface:

**Pull request merged**